### PR TITLE
[SPARK-14061][SQL] implement CreateMap

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

As we have `CreateArray` and `CreateStruct`, we should also have `CreateMap`.  This PR adds the `CreateMap` expression, and the DataFrame API, and python API.
## How was this patch tested?

various new tests.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @rxin

File: python/pyspark/sql/functions.py [hunk 1]
```diff
@@ -1498,6 +1498,25 @@ def translate(srcCol, matching, replace):
 
 # ---------------------- Collection functions ------------------------------
 
+@since(2.0)
+def map(*cols):
```
Comment (rxin) [inline python/pyspark/sql/functions.py]: this actually conflicts with the bulit in python map function

**Comment by cloud-fan** [comment]
retest this please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ComplexTypeSuite.scala [hunk 2]
```diff
@@ -134,6 +134,46 @@ class ComplexTypeSuite extends SparkFunSuite with ExpressionEvalHelper {
     checkEvaluation(CreateArray(Literal.create(null, IntegerType) :: Nil), null :: Nil)
   }
 
+  test("CreateMap") {
+    def interlace(keys: Seq[Literal], values: Seq[Literal]): Seq[Literal] = {
+      keys.zip(values).flatMap { case (k, v) => Seq(k, v) }.map(Literal(_))
```
Comment (liancheng) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ComplexTypeSuite.scala]: Are we creating `Literal`s of `Literal`s here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala [hunk 3]
```diff
@@ -160,6 +160,9 @@ object HiveTypeCoercion {
     })
   }
 
+  private def isSameType(exprs: Seq[Expression]): Boolean =
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: `haveSameType` might be better?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ComplexTypeSuite.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ComplexTypeSuite.scala]: oh, it's a mistake, good catch!

**Comment by liancheng** [comment]
LGTM pending Jenkins.

**Pull request closed without merging**